src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeExceptionHelpers.cs

```diff
@@ -165,102 +166,99 @@ public static void RuntimeFailFast(RhFailFastReason reason, Exception? exception
-                }
-
-                FailFast(failFastMessage, exception, reason, pExAddress, pExContext);
+                FailFast(message: null, exception, reason, pExAddress, pExContext);
             }
             catch
             {
-                // Returning from this callback will cause the runtime to FailFast without involving the class
-                // library.
+                // Returning from this callback will cause the runtime to FailFast without involving the class library.
             }
         }
 
+        private static ulong s_crashingThreadId;
+
         [DoesNotReturn]
-        internal static unsafe void FailFast(string message, Exception? exception, RhFailFastReason reason, IntPtr pExAddress, IntPtr pExContext)
+        internal static unsafe void FailFast(string? message, Exception? exception, RhFailFastReason reason, IntPtr pExAddress, IntPtr pExContext)
         {
-            // If this a recursive call to FailFast, avoid all unnecessary and complex activity the second time around to avoid the recursion
-            // that got us here the first time (Some judgement is required as to what activity is "unnecessary and complex".)
-            bool minimalFailFast = InFailFast.Value || (exception == PreallocatedOutOfMemoryException.Instance);
-            InFailFast.Value = true;
-
-            CrashInfo crashInfo = new();
-            crashInfo.Open(reason, message);
+            IntPtr triageBufferAddress = IntPtr.Zero;
+            int triageBufferSize = 0;
+            int errorCode = 0;
 
-            if (!minimalFailFast)
+            ulong currentThreadId = Thread.CurrentOSThreadId;
+            ulong previousThreadId = Interlocked.CompareExchange(ref s_crashingThreadId, currentThreadId, 0);
+            if (previousThreadId == 0)
             {
-                string prefix;
-                string outputMessage;
-                if (exception != null)
-                {
-                    prefix = "Unhandled Exception: ";
-                    outputMessage = exception.ToString();
-                }
-                else
+                CrashInfo crashInfo = new();
+                crashInfo.Open(reason, s_crashingThreadId, message);
+
+                bool minimalFailFast = (exception == PreallocatedOutOfMemoryException.Instance);
+                if (!minimalFailFast)
                 {
-                    prefix = "Process terminated. ";
-                    outputMessage = message;
+                    string prefix;
+                    string outputMessage;
+                    if (exception != null)
+                    {
+                        prefix = "Unhandled Exception: ";
+                        outputMessage = exception.ToString();
+                    }
+                    else
+                    {
+                        prefix = "Process terminated. ";
+                        message ??= string.Format("Runtime-generated FailFast: ({0}): {1}",
```

Comment:
`Environment.FailFast(null);` is going to print "Process terminated. Runtime-generated FailFast: (EnvironmentFailFast): Unknown reason." with this change. That is too much random words.

I think we should change this to just `message ??= GetStringForFailFastReason(reason)`, and maybe also make `GetStringForFailFastReason` to return something better than just "Unknown reason." for EnvironmentFailFast reason.

Reply:
Now I'm thinking that the `message ??= GetStringForFailFastReason(reason)` should be above the if minimalFailFast condition (OOM) and the resulting `message` passed to CrashInfo.Open(), but I'm not sure how the literal strings that GetStringFor... returns would be allocated.

Reply:
Literal strings are created at compile time in native AOT. Returning literal string does not incur allocations in native AOT.